### DISCUSSION TITLE
CDN distribution

### DISCUSSION THREAD

**Discussion by varvello** [robertknight/tesseract-wasm#92] [General]
Hi Robert,
Isn't there a CDN distribution of tesseract-wasm? I'm trying to use it in a browser and It should be simpler than using npm.

Cheers
Davide

**Comment by robertknight** [comment]
In short, no. While the files are available on general npm CDNs like [unpkg](https://unpkg.com/browse/tesseract-wasm@0.10.0/) or [jsDelivr](https://www.jsdelivr.com/package/npm/tesseract-wasm), I believe you will encounter issues trying to load this library from them, because of the browser restriction that Web Workers must be served from the same origin as the current page.

[Tesseract.js](https://github.com/naptha/tesseract.js) meanwhile does have a CDN distribution, and it looks like they use a [workaround](https://github.com/naptha/tesseract.js/blob/master/src/worker/browser/spawnWorker.js#L2) for the same-origin restriction for workers. Perhaps tesseract-wasm could integrate something similar in future.

**Reply by varvello** [reply]
Thanks Robert,
Yep, I successfully tested Tesseract.js using the CDN and I'd like to do some benchmarks with tesseract-wasm, but I rarely use Node and I haven't been able to run tesseract-wasm in the browser so far

**Comment by robertknight** [comment]
It turns out it is possible to load tesseract-wasm from jsDelivr using a similar approach to Tesseract.js, with a workaround for an issue with URLs.

Here is a demo HTML example, which you will need to serve via some local HTTP server (or upload to an HTTP URL):

```html
<html>
  <body>
    <script type="module">
      import { OCRClient } from 'https://cdn.jsdelivr.net/npm/tesseract-wasm@0.10.0/+esm';

      async function runOCR() {
        // Fetch document image and decode it into an ImageBitmap.
        const imageResponse = await fetch('https://i.imgur.com/YfEZa63.jpeg');
        const imageBlob = await imageResponse.blob();
        const image = await createImageBitmap(imageBlob);

        const wasmResponse = await fetch(
          'https://cdn.jsdelivr.net/npm/tesseract-wasm@0.10.0/dist/tesseract-core.wasm'
        );
        const wasmBinary = await wasmResponse.arrayBuffer();

        const realWorkerURL =
          'https://cdn.jsdelivr.net/npm/tesseract-wasm@0.10.0/dist/tesseract-worker.js';
        const workerSource = new Blob([`
          // Make URL constructor a no-op to work around an error when emscripten
          // tries to determine the WASM URL (which is not essential here).
          globalThis.URL = function () {};
          importScripts("${realWorkerURL}");
        `], {
          type: 'application/javascript',
        });
        const workerURL = URL.createObjectURL(workerSource);

        // Initialize the OCR engine. This will start a Web Worker to do the
        // work in the background.
        const ocr = new OCRClient({
          workerURL,
          wasmBinary,
        });

        try {
          // Load the appropriate OCR training data for the image(s) we want to
          // process.
          await ocr.loadModel(
            'https://raw.githubusercontent.com/tesseract-ocr/tessdata_fast/main/eng.traineddata'
          );

          await ocr.loadImage(image);

          // Perform text recognition and return text in reading order.
          const text = await ocr.getText();

          console.log('OCR text: ', text);
        } finally {
          // Once all OCR-ing has been done, shut down the Web Worker and free up
          // resources.
          ocr.destroy();
        }
      }

      runOCR();
    </script>
  </body>
</html>
```

**Reply by varvello** [reply]
That's really really great, Robert!! I'm going to try your solution and feedback you
Thank you very much

**Reply by varvello** [reply]
Hi Robert, it's working!!

A couple of raw numbers on a m2 mac: using the image https://i.imgur.com/YfEZa63.jpeg text recognition with tesseract-wasm is about **200msec** instead with tesseract.js is about **300msec**

**Reply by varvello** [reply]
Raw benchmarks with some phones:

- Redmi note 7: tesseract.js ~2500 msec - tesseract-wasm 1300 msec
- Samsung galaxy m21:  tesseract.js 2200 msec - tesseract-wasm  1200 msec
- iPhoneXR:  tesseract.js 800 msec - tesseract-wasm  600 msec

**Reply by robertknight** [reply]
Interesting. I wonder what Tesseract.js is doing differently. The main advantage of this project originally was that it used WASM SIMD, but since [v3](https://github.com/naptha/tesseract.js?tab=readme-ov-file#major-changes-in-v3) Tesseract.js does as well.

**Reply by varvello** [reply]
I've no clue, my test code using Tesseract.js is really simple

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <script src="https://cdn.jsdelivr.net/npm/tesseract.js@5/dist/tesseract.min.js"></script>
    <title></title>
  </head>
  <body>
    <div class="container">
      <div class="containerTitle">
        <h1>Demo TesseractJS OCR</h1>
      </div>
    </div>
    <script>
      async function performOCR() {
        const worker = await Tesseract.createWorker("eng");
        let startTime = new Date().getTime();
        const {data: { text }, } = await worker.recognize("./images/YfEZa63.jpeg");
        let endTime = new Date().getTime();
        console.log("JS OCR text: ", text);
        console.log(endTime - startTime);
      }

      performOCR();
    </script>
  </body>
</html>

```
